docs: add History Expiry wiki article (#499)
eth-protocol-fellows/protocol-studies: docs/wiki/research/history-expiry/history-expiry.md

# History Expiry In Ethereum

> **Recommended pre-reading**
> - [Ethereum Node Architecture](/wiki/EL/el-specs.md)
> - [Execution Layer Specification](/wiki/EL/el-specs.md)
> - [DevP2P protocol](/wiki/EL/devp2p.md)
> - [Data structures and encoding](/wiki/EL/data-structures.md)
> - [EIP-4444: Bound Historical Data in Execution Clients](https://eips.ethereum.org/EIPS/eip-4444)
> - [EIP-7643: History Accumulator for Pre-PoS Data](https://eips.ethereum.org/EIPS/eip-7643)

History expiry is the idea that nodes should not be required to store historical data forever. Every Ethereum full node stores two categories of data. State is the current account, contract storage, and code. History is past block headers, bodies, and receipts.

> Up until recently, full nodes were expected to store and serve all historical data, over the peer-to-peer network, even if the data is not needed to validate a block. That data has grown with every block since genesis. Today, running a full node takes well over a terabyte of disk space, and both client load and sync times keep increasing even as the chain's capacity remains the same.

Reviewer: I would emphasize the sync strategy that became default. The missing historical data can be a sensitive topic because without the history, it's not possible to fully validate the chain from genesis to the tip (which is default in Bitcoin for example). But thanks to weak subjectivity checkpoints and snap syncing, we can still sync securely and opt in for history/full validation

Reviewer 2: Good point, added a paragraph explaining that Ethereum's default sync strategy already does not validate from genesis. Mentioned snap sync and weak subjectivity checkpoints, and clarified that full historical validation is still possible but optional.

To solve this, [EIP-4444](https://eips.ethereum.org/EIPS/eip-4444) proposes that nodes may prune historical data and receipts older than a set threshold and stop serving them over the peer-to-peer network. Once a client has synced to the tip of the chain, historical data is only retrieved when requested explicitly over the JSON-RPC or when a peer attempts to sync. For this to work, the peer-to-peer protocol itself needed to change.

## DevP2P Changes

> Under `eth/68`, nodes assumed every peer stored the full chain from genesis. A node pruning old history while still advertising itself on `eth/68` would break that assumption and disrupt sync for peers requesting old blocks. [EIP-7642](https://eips.ethereum.org/EIPS/eip-7642) introduces `eth/69`, which removes this assumption. Prior to `eth/69`, when two nodes connect, they exchange a Status message containing the network ID, genesis hash, fork ID, and the hash of the node's latest block, but now the status handshake includes two new fields, `earliestBlock` and `latestBlock` that store the block range.

Reviewer: ```suggestion
Under `eth/68` and older `eth` protocols, nodes assumed every peer stored the full chain from genesis. A node pruning old history while still advertising itself on `eth/68` would break that assumption and disrupt sync for peers requesting old blocks. [EIP-7642](https://eips.ethereum.org/EIPS/eip-7642) introduces `eth/69`, which removes this assumption. Prior to `eth/69`, when two nodes connect, they exchange a Status message containing the network ID, genesis hash, fork ID, and the hash of the node's latest block, but now the status handshake includes two new fields, `earliestBlock` and `latestBlock` that store the block range.
```

Reviewer 2: Added 'and older `eth` protocols' locally

```python
    # Old eth/68 Status
    # [version, networkid, td, blockhash, genesis, forkid]

    # New eth/69 Status
    # Now includes earliestBlock and latestBlockHash.
    # td (totalDifficulty) is removed as it is useless since the merge.
    # [version, networkid, genesis, forkid, earliestBlock, latestBlock, latestBlockHash]

    # BlockRangeUpdate message, sent when a node's range changes.
    # Sent at most once per epoch (32 blocks).
    # [earliestBlock, latestBlock, latestBlockHash]
```

`eth/69` also adds a new message, BlockRangeUpdate. As a node prunes more data or downloads more history, it sends this message to its connected peers so they can update their view of what blocks that node can serve. This only needs to be sent once per epoch.

> `eth/69`'s linear range works for Phase 1 where nodes either hold pre-merge data or they don't. For Phase 2, where nodes may hold non-contiguous slices of history, proposals like [EIP-7801](https://eips.ethereum.org/EIPS/eip-7801), still in draft, introduce a bitmask-based subprotocol called `etha` that lets nodes advertise exactly which segments of the chain they store. While the eth protocol continues to handle live chain operations like block propagation, transaction gossip, and syncing to the tip, the `etha` subprotocol is dedicated entirely to serving historical data. This means historical block requests will no longer travel over the same channel as live chain, so a node that is looking for old blocks queries peers over etha, and nodes that do not support history sharding are never bothered with those requests again.

Reviewer: I would specify this as it's the current state -  that the Merge is cut off for historical data as a clear historical point, not a moving target. 
```suggestion
`eth/69`'s linear range works for Phase 1 where nodes either hold pre-Merge (old PoW chain) data or they don't. For Phase 2, where nodes may hold non-contiguous slices of history, proposals like [EIP-7801](https://eips.ethereum.org/EIPS/eip-7801), introduce a bitmask-based subprotocol called `etha` that lets nodes advertise exactly which segments of the chain they store. While the eth protocol continues to handle live chain operations like block propagation, transaction gossip, and syncing to the tip, the `etha` subprotocol is dedicated entirely to serving historical data. This means historical block requests will no longer travel over the same channel as live chain, so a node that is looking for old blocks queries peers over etha, and nodes that do not support history sharding are never bothered with those requests again.
```

Reviewer 2: Changed to 'pre-Merge (old PoW chain)' and removed 'still in draft

**Window: Blocks 0 — 1,064,960**

|              | Span 0 | Span 1 | Span 2 | Span 3 | Span 4 | Span 5 | Span 6 | Span 7 | Span 8 | Span 9 |
|--------------|--------|--------|--------|--------|--------|--------|--------|--------|--------|--------|
| Node A       |   X    |        |        |        |        |        |        |        |        |        |
| Node B       |        |        |        |   X    |        |        |        |        |        |        |
| Node C       |        |        |        |        |        |        |        |   X    |        |        |
| Node D       |        |   X    |        |        |        |        |        |        |        |        |
| Node E       |        |        |        |        |        |   X    |        |        |        |        |
| Node F       |        |        |        |        |        |        |        |        |        |   X    |

> **Window: Blocks 1,064,960 — 2,129,920**

Reviewer: Having 2 tables to demonstrate this feels a bit redundant

Reviewer 2: I agree, deleted the second table

$$\text{epoch} = \left\lfloor \frac{500{,}000}{8{,}192} \right\rfloor = 61$$

$$\text{position within epoch} = 500{,}000 \mod 8{,}192 = 888$$

The Merkle proof from `HeaderRecord` at index 888 to `epoch_root_61` requires

$$\log_2(8192) = 13 \text{ sibling hashes}$$

The Merkle proof from `epoch_root_61` to the full accumulator root requires

$$\log_2(2048) = 11 \text{ sibling hashes}$$

The total proof size is

$$13 + 11 = 24 \text{ hashes} \times 32 \text{ bytes} = 768 \text{ bytes}$$

to prove any pre-merge block is canonical.

This is what the [Portal Network](https://www.ethportal.net/) uses. When a node requests a historical block from the network, the response includes the block data plus a Merkle proof against the accumulator.

### Distribution

ERA1 files follow the naming convention `<network>-<epoch>-<hexroot>.era1`, for example `mainnet-00000-5ec1ffb8.era1` for the first 8,192 blocks on mainnet. The hex portion is a truncated accumulator root, so the filename itself is a quick integrity check. The BlockIndex at the end of the file stores relative offsets to each block tuple, making random access by block number possible without scanning the entire file.

> The [eth-clients/history-endpoints](https://github.com/eth-clients/history-endpoints) registry maintains a community list of providers serving ERA1 and ERA files over HTTP and torrents. Providers like [ethPandaOps](https://ethpandaops.io/data/history/) host the full set of mainnet ERA1 files with SHA256 checksums for verification. Files can also be shared over BitTorrent. The goal is that no single provider is required and the data remains available through multiple independent channels.

Reviewer: why not add the whole magnet link
```suggestion
The [eth-clients/history-endpoints](https://github.com/eth-clients/history-endpoints) registry maintains a community list of providers serving ERA1 and ERA files over HTTP and torrents. Providers like [ethPandaOps](https://ethpandaops.io/data/history/) host the full set of mainnet ERA1 files with SHA256 checksums for verification. Files can also be shared over [BitTorrent](magnet:?xt=urn:btih:edcc7c112bae520e3226065a61817d3575904e0d&dn=EthereumMainnetPreMergeEraFiles&xl=458498121702&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337%2Fannounce&tr=udp%3A%2F%2Fopen.tracker.cl%3A1337%2Fannounce&tr=udp%3A%2F%2Fbt1.archive.org%3A6969%2Fannounce). The goal is that no single provider is required and the data remains available through multiple independent channels.
```

Reviewer 2: Fixed, thanks for pointing out!

Client support is already in place. [Geth](https://geth.ethereum.org/docs/fundamentals/downloadera), [Nimbus](https://nimbus.guide/era-store.html), [Besu](https://besu.hyperledger.org/public-networks/how-to/era1-file-full-sync), and [Reth](https://reth.rs/docs/reth_era/index.html) all support ERA1 imports. Each etha span of 106,496 blocks is exactly 13 ERA1 files (13 × 8,192 = 106,496), so a node's storage boundaries under etha map directly to whole ERA1 files.

> ## Portal Network

Reviewer: Portal is currently sort of discontinued and not developed. Either way, it would deserve it's own page. I would maybe keep it short here and dive into Portal on a subpage

Reviewer 2: Trimmed the Portal Network section and flagged stalled status. Will write a dedicated Portal Network page separately.
